output/elasticsearch/config_test.go

```diff
@@ -26,6 +28,78 @@ func TestInvalidConfig(t *testing.T) {
 	// that an invalid outcome is possible.
 }
 
+func TestWatcherPeriodWithNoEvents(t *testing.T) {
+	report := func(state WatchState, s string) {
+		if state == WatchDegraded {
+			t.Logf("got: %s, should not have failed", s)
+			t.Fail()
+		}
+	}
+	watcher := ESHeathWatcher{failureInterval: time.Second, reporter: report, waitInterval: time.Millisecond * 100}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	go watcher.Watch(ctx)
+	watcher.Success()
+	// should not report failure if it's just been a long period without any events
+	time.Sleep(time.Second * 1)
+}
+
+func TestWatcherWithFail(t *testing.T) {
+	gotFail := false
+	reportFail := func(state WatchState, s string) {
+		t.Logf("got: %s", s)
+		if state == WatchDegraded {
+			gotFail = true
+		}
+
+	}
+	watcher := ESHeathWatcher{failureInterval: time.Millisecond * 400, reporter: reportFail, waitInterval: time.Millisecond * 100}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	go watcher.Watch(ctx)
+
+	watcher.Success()
+	time.Sleep(time.Millisecond * 100)
+	watcher.Fail("simulated failure")
+	// should fail
+	time.Sleep(time.Millisecond * 600)
```

Comment:
The sleeps in these tests are some heavy costs to impose on every unit test run. Can we instead make the time callback a parameter of the watcher? Typical use could pass in `time.Now` (or maybe that could be the default) while unit tests could pass in a mocked timer so it can be tested deterministically without any delay.

Reply:
So, I tinkered with this for a bit, but after I started running into race issues between the tests and the main watch loop, I decided it would be easier to just shift things around to make the sleeps faster, which is _hopefully_ good enough.

Reply:
Mmm I see what you mean but it makes me sad. Probably the Right thing to do here is still to pass in a mockable helper to generate timer channels or something so we can make it deterministic, but that would require a lot of revisions for something that already mostly works. Maybe we can do a more robust generic helper when we have more components with health to report; I'll approve this one for now :-)

Reply:
Yah, agreed, I'm not a fan of it either, and I experimented with adding a few helpers/callbacks, but it just felt like the added code complexity just wasn't worth making the tests fully deterministic, particularly after shortening the sleeps in the test.